### ISSUE TITLE
InvalidOperationException: Collection was modified

### ISSUE THREAD

**Issue by db-ks** [opentap/opentap#762]
**Describe the bug**

I intermitently get an error when I try to run a plan.

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04
 - OpenTAP Version 9.19.0-beta.19

1. Run plan

**Actual behavior**

```json
{
  "Logs": [
    {
      "Source": "AssemblyData",
      "Timestamp": 0,
      "Message": "Loaded OpenTap.Package. [1.05 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "No settings file exists for DutSettings. A new instance with default values has been created. [1.49 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "No settings file exists for InstrumentSettings. A new instance with default values has been created. [222 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "No settings file exists for ConnectionSettings. A new instance with default values has been created. [118 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "No settings file exists for ResultSettings. A new instance with default values has been created. [143 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.Extensions.DependencyModel, Version=2.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.Extensions.DependencyModel.2.1.0.0/Microsoft.Extensions.DependencyModel.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "REST-API",
      "Timestamp": 0,
      "Message": "Built WebHost [316 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for System.IO.Pipelines, Version=4.0.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/System.IO.Pipelines.4.0.0.1/System.IO.Pipelines.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.AspNetCore.StaticFiles, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.AspNetCore.StaticFiles.2.2.0.0/Microsoft.AspNetCore.StaticFiles.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.IdentityModel.Protocols.OpenIdConnect, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.IdentityModel.Protocols.OpenIdConnect.6.16.0.0/Microsoft.IdentityModel.Protocols.OpenIdConnect.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.IdentityModel.Tokens, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.IdentityModel.Tokens.6.16.0.0/Microsoft.IdentityModel.Tokens.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.IdentityModel.Protocols, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.IdentityModel.Protocols.6.16.0.0/Microsoft.IdentityModel.Protocols.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.IdentityModel.Logging, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.IdentityModel.Logging.6.16.0.0/Microsoft.IdentityModel.Logging.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "EESOFv9",
      "Timestamp": 0,
      "Message": "Checked out: KS8106A 1.0 [194 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "License",
      "Timestamp": 0,
      "Message": "License resolved: KS8106A|TAP_RestApi [282 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.AspNetCore.Http.Connections.Common, Version=1.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.AspNetCore.Http.Connections.Common.1.1.0.0/Microsoft.AspNetCore.Http.Connections.Common.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Namotion.Reflection, Version=2.0.10.0, Culture=neutral, PublicKeyToken=c2f9c3bdfae56102 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Namotion.Reflection.2.0.10.0/Namotion.Reflection.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.Extensions.FileProviders.Embedded, Version=1.0.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.Extensions.FileProviders.Embedded.1.0.1.0/Microsoft.Extensions.FileProviders.Embedded.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "AssemblyData",
      "Timestamp": 0,
      "Message": "Loaded OpenTap.Plugins.RestApi.Shared. [5.58 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.Net.Http.Headers, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.Net.Http.Headers.2.2.0.0/Microsoft.Net.Http.Headers.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.AspNetCore.JsonPatch, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.AspNetCore.JsonPatch.2.2.0.0/Microsoft.AspNetCore.JsonPatch.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for NSwag.Annotations, Version=13.15.10.0, Culture=neutral, PublicKeyToken=c2d88086e098d109 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/NSwag.Annotations.13.15.10.0/NSwag.Annotations.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "REST-API",
      "Timestamp": 0,
      "Message": "Swagger page: http://127.0.0.1:7038/swagger",
      "Level": 30,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for System.IdentityModel.Tokens.Jwt, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/System.IdentityModel.Tokens.Jwt.6.16.0.0/System.IdentityModel.Tokens.Jwt.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "REST-API",
      "Timestamp": 0,
      "Message": "Session started!",
      "Level": 30,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.AspNetCore.Http.Extensions, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.AspNetCore.Http.Extensions.2.2.0.0/Microsoft.AspNetCore.Http.Extensions.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.AspNetCore.WebUtilities, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.AspNetCore.WebUtilities.2.2.0.0/Microsoft.AspNetCore.WebUtilities.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Microsoft.IdentityModel.JsonWebTokens, Version=6.16.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Microsoft.IdentityModel.JsonWebTokens.6.16.0.0/Microsoft.IdentityModel.JsonWebTokens.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "AuthenticationSettings loaded from /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Settings/AuthenticationSettings.xml [15.9 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "License",
      "Timestamp": 0,
      "Message": "License resolved: KS8106A|TAP_RestApi [312 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "CampaignSettings loaded from /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Settings/Campaign.xml [9.00 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "AssemblyData",
      "Timestamp": 0,
      "Message": "Loaded Keysight.OpenTap.Plugins.ResultServiceClient. [727 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "License",
      "Timestamp": 0,
      "Message": "License resolved: KS8106A|TAP_RestApi [173 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Keysight.OpenTap.ResultService.Integration, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Keysight.OpenTap.ResultService.Integration.1.0.0.0/Keysight.OpenTap.ResultService.Integration.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Resolver",
      "Timestamp": 0,
      "Message": "Found match for Keysight.OpenTap.Shared.Defense, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null in /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Dependencies/Keysight.OpenTap.Shared.Defense.1.0.0.0/Keysight.OpenTap.Shared.Defense.dll",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "ResultSettings loaded from /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Settings/Results.xml [15.3 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "RPC Base",
      "Timestamp": 0,
      "Message": "Running TestPlan MyCampaign-kuri4l",
      "Level": 30,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "-----------------------------------------------------------------",
      "Level": 30,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "Starting TestPlan 'MyCampaign-kuri4l' on 08/18/2022 11:20:27, 2 of 2 TestSteps enabled.",
      "Level": 30,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "EngineSettings loaded from /tmp/opentap-images/8A5847F20E5825594723F181771F9CCFE695A993/Settings/Engine.xml [17.8 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "CLI",
      "Timestamp": 0,
      "Message": "Executing CLI action: package \\ check-updates",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Settings",
      "Timestamp": 0,
      "Message": "No settings file exists for PackageManagerSettings. A new instance with default values has been created. [627 us]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "CLI",
      "Timestamp": 0,
      "Message": "CLI action returned exit code: 0 [4.55 ms]",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "Campaign",
      "Timestamp": 0,
      "Message": "Repository 'http://packages.opentap.keysight.com/4.0/query' response NotFound. Message:",
      "Level": 10,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "TestPlan aborted.",
      "Level": 20,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "Collection was modified",
      "Level": 10,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "InvalidOperationException: Collection was modified",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TypeData.GetDerivedTypes(ITypeData baseType)",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TypeData.GetDerivedTypes[BaseType]()",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TestPlanRunMonitors.GetCurrent()",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TestPlan.OpenInternal(TestPlanRun run, Boolean isOpen, List`1 steps)",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TestPlan.DoExecute(IEnumerable`1 resultListeners, IEnumerable`1 metaDataParameters, HashSet`1 stepsOverride)",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "Exception caught at:",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TestPlanRun DoExecute(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep])",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at OpenTap.TestPlanRun executeInContext(System.Collections.Generic.IEnumerable`1[OpenTap.IResultListener], System.Collections.Generic.IEnumerable`1[OpenTap.ResultParameter], System.Collections.Generic.HashSet`1[OpenTap.ITestStep])",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at Void <ExecutePlan>b__0()",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "at Void Process()",
      "Level": 40,
      "DurationNS": 0
    },
    {
      "Source": "TestPlan",
      "Timestamp": 0,
      "Message": "...",
      "Level": 40,
      "DurationNS": 0
    }
  ],
  "Offset": 0,
  "FilteredCount": 58,
  "TotalCount": {
    "10": 2,
    "20": 1,
    "30": 5,
    "40": 50
  }
}
```

**Expected behavior**

The plan would start without any problems.

**Comment by rmadsen-ks** [comment]
This was closed by #768

**Comment by AsgerIversen** [comment]
This change is part of OpenTAP version 9.19.0-beta.27+250f38ab or later.
